Add tests for Service Request Create

yash-learner opened the pull request:
## Proposed Changes

Fixes #14296
- Change 1
- Change 2
- Additional context if needed

Tagging: @ohcnetwork/care-fe-code-reviewers

## Merge Checklist

- [ ] Add specs that demonstrate the bug or test the new feature.
- [ ] Update [product documentation](https://docs.ohc.network).
- [ ] Ensure that UI text is placed in I18n files.
- [ ] Prepare a screenshot or demo video for the changelog entry and attach it to the issue.
- [ ] Request peer reviews.
- [ ] Complete QA on mobile devices.
- [ ] Complete QA on desktop devices.
- [ ] Add or update Playwright tests for related changes


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Service request details now show body site and patient instruction when present.

* **Tests**
  * Added end-to-end tests covering service request creation (required and all fields) and detail-view validation.
  * Added test helpers, shared test constants, and UI interaction utilities to simplify test flows.
  * Added a setup flow to capture encounter metadata for tests.

* **Chores**
  * Restructured test setup projects and adjusted test execution configuration.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

yash-learner committed:
wip: Add test for activity definition form

- Add two cases
- Add api-context util
- Add helper to get facility id and category id

yash-learner committed:
add alias

yash-learner committed:
refactor: update facility and resource category retrieval in tests

- Change `getFacilityAndCategory` to return a `FacilityRead` object instead of `facilityId`.
- Update helper functions to fetch the last facility and resource category because API returns in desc order
- Adjust test cases to use the new structure for accessing facility and resource category data.

yash-learner committed:
refactor: introduce selectFromCommand helper for streamlined UI interactions in tests

- Add `selectFromCommand` function to encapsulate logic for selecting options from command inputs.
- Refactor activity definition form test to utilize the new helper for improved readability and maintainability.

yash-learner committed:
Enhance activity definition form tests with comprehensive field validation

- Update `activity-definition-form.spec.ts` to include a new test for filling out the form with all fields and submitting.
- Refactor `selectFromCommand` and add `selectFromLocationMultiSelect` helper functions to improve UI interaction handling.
- Implement defensive checks to close any open popovers before interactions to ensure stability during tests.

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
fix: update API base URL in Playwright configuration

- Change `apiBaseURL` from `PLAYWRIGHT_API_BASE_URL` to `REACT_CARE_API_URL`

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
wip: Organise specs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Refactor: reorganize test helpers and setup for facility data

- Remove deprecated `helpers.ts` and introduce `helpers/helpers.ts` for improved modularity.
- Implement `loadFacility` and `getResourceCategory` functions to streamline facility and resource category retrieval.
- Add `facility.setup.ts` to prepare and cache facility data for tests, enhancing test reliability.
- Update `activity-definition-form.spec.ts` to utilize new helper functions for better test structure.

yash-learner committed:
Rename file and add cases for existing category

yash-learner committed:
Remove explicit timeouts

yash-learner committed:
Fix typo in file name

yash-learner committed:
Improve naming a bit

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
WIP: Make changes as per suggestions from Bodhi

- Update categoryList
- Remove API files
- use facility ID

yash-learner committed:
Update Playwright configuration and refactor activity definition tests

- Increased retry count for local in playwright.config.ts
- Refactored activity definition tests to use facility ID and improved selector functions
- Removed unused imports and streamlined test setup
- Introduced new helper functions for selecting from requirements and valuesets and category picker

yash-learner committed:
Address some copilot comments

yash-learner committed:
Add test for creating nested category in activity definitions

yash-learner committed:
Add edit case and other refactors

yash-learner committed:
Remove the explicit timeout, to fix CI - Let PW handle it

https://github.com/ohcnetwork/care_fe/actions/runs/18927252776/job/54037864162?pr=14108

yash-learner committed:
Add delete AD test case and other changes

yash-learner committed:
Revert change made to user schedule creation

yash-learner committed:
Refactor and be specific with selectors

yash-learner committed:
Add Playwright tests for activity definitions list functionality

- Fix back button for view page

yash-learner committed:
Fix some flakyness

yash-learner committed:
CI tests can run in parallel shards so create test can run before list

yash-learner committed:
Tiny refactor and make a visible check for count so that it is not flaky in local

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Fix import paths after merging the develop

yash-learner committed:
Add test to verify content in activity definition table row

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Update based on new changes in the UI by recent PR merge

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Revert changes made to the pw config

yash-learner committed:
wip: Add the create service request spec

- Add create service request flow spec
- Add helper for picking from RCDP
- Add setup to store and reuse patient id and encounter id

yash-learner committed:
Refactor create service request test

- Add reusable selector to automatically from tab or menuitem
- Add resuable helper to create service requests

yash-learner committed:
Add lab location setup and metadata retrieval utilities

- Implement lab location setup test to find and save Bio-Chemistry lab location metadata.
- Create utility functions to retrieve lab location metadata and IDs from saved JSON file.
- Ensure automatic setup execution if metadata is missing or invalid.

nihal467 committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Update text based on change in PR#14308

-  Add video retention on first retry

yash-learner committed:
Revert change to retries

yash-learner committed:
Update text for other test as well

yash-learner committed:
Merge branch 'add-tests-for-labs' into issues/14296/add-tests-for-service-request

yash-learner committed:
Enhance Playwright configuration by adding specific setup projects
for authentication, facility, encounter, and location tests.
Update dependencies to ensure proper execution order.

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Randomise the status and classification

Renama categoryName to resourceCategotyName

yash-learner committed:
Address review comments

- Separate edit and delete files
- Separate test cases for verify details
- Helper file to generate relevant data for specs
- Randomise
- More

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Check slug auto generation and some refactor in edit

yash-learner committed:
Quick test to test timeout

- Add some AD to check data for all status and classification

yash-learner committed:
Refactor activity definition list tests to use parameterized status and classification filtering

yash-learner committed:
Moving list specs to a separate follow up PR

yash-learner committed:
Add case for validation errors

yash-learner committed:
Remove duplicate visibility check for slug required error in activity definition creation tests

yash-learner committed:
Move to faker arrayElement from faker number int for SD, OD, Location, CID, DR

- Update  the tests

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Add copilot suggestions

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Handle multi text match

yash-learner committed:
Push missed out changes

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Rename folder name

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Use clearFilter

yash-learner committed:
Merge branch 'add-tests-for-labs' of https://github.com/ohcnetwork/care_fe into add-tests-for-labs

yash-learner committed:
Refactor slug generation: move generateExpectedSlug to utils and update activity definition tests to use the new import. Clean up LOCATIONS array in activityDefinition helper.

yash-learner committed:
wip: Address review comments

Refactor activity definition tests: streamline creation logic by removing resourceCategoryName parameter,
update slug generation to use expectedSlug utility, and
enhance test data generation for improved clarity and maintainability.

yash-learner committed:
Merge branch 'develop' into add-tests-for-labs

yash-learner committed:
Update validation error case to utilize getFieldErrorMessage

yash-learner committed:
Move ad helper from helper folder to under activityDefinition folder

yash-learner committed:
Refactor activity definition data generation and update related tests

- Updated `generateActivityDefinitionData` to include optional fields and improved structure.
- Refactored `createActivityDefinition` to utilize the new data generation function.
- Adjusted Playwright tests for activity definition creation, editing, and deletion to align with the updated data structure and navigation paths.
- Removed unused `clearFilter` helper function from UI tests.

yash-learner committed:
Fix path

yash-learner committed:
Refactor Playwright tests to use getCardByTitle helper for card element selection

- Replaced direct locator usage with the new `getCardByTitle` function in both activity definition create and edit tests for improved readability and maintainability.
- Added `getCardByTitle` function in the UI helper to streamline card selection by title.

yash-learner committed:
Fix the js doc

yash-learner committed:
Assert slug instead of fill in CAD helper

yash-learner committed:
Tiny change: Update test descriptions a bit

yash-learner committed:
Address a coderabbit nitpick

- Simplified the return statement in `createActivityDefinition` by directly returning the `data` object, removing unnecessary destructuring of fields.

yash-learner committed:
Address coderabbit suggestion

yash-learner committed:
Fill slug in edit

yash-learner committed:
Move ui.ts, utils.ts from helpers to helper

yash-learner committed:
Merge branch 'add-tests-for-labs' into issues/14296/add-tests-for-service-request

yash-learner committed:
wip: Refactor SR specs based on learning from other Pr's

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request

yash-learner committed:
Add body site and patient instruction display to ServiceRequestDetails

yash-learner committed:
wip: Add requestor field and selection helper for service requests;
introduce known users constant

yash-learner committed:
Some refactor

- Add static list for priority
- Some more changes

yash-learner committed:
Remove labLocationSetup from this PR as I am

making list specs and lab activity into separate PR's
This is to reduce turn around time and for east reviews

yash-learner committed:
Clean up

yash-learner committed:
Revert some changes

yash-learner committed:
Clean up

yash-learner committed:
Check the requestor name

-  store the name
-  Some refactors

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request

nihal467 commented on tests/facility/encounter/serviceRequests/ServiceRequestCreate.spec.ts:
```diff
@@ -0,0 +1,102 @@
+
+    await expect(
+      page
+        .locator(".text-gray-700")
```
- try to find a better selector

yash-learner replied:
I remember, I couldn't find anything sematic for some of these so went with tailwind classes


Will look again, will try to fix it

yash-learner replied:
@nihal467 

Should I just use a div locator here ?


because we have moved away from `data-test-id` and this piece code does not use any shadcn component to use data-slot


```
    <div className="bg-gray-100 rounded-lg border border-gray-200">
      <div className="py-3 flex items-center justify-between">
        <div className="text-sm text-gray-600">
          <div className="font-semibold text-gray-600 text-xl flex items-center gap-2">
            <div className="h-8 w-1.5 bg-gray-400 rounded-r-sm" />
            <div className="flex items-center gap-2 text-gray-700">
              {activityDefinition.title}
            </div>
          </div>
  ```

nihal467 reviewed (changes requested):
- Look into the Copilot suggestion for better Tailwind
- fix the conflict

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request

yash-learner committed:
Fix conflicts and review comments

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request

yash-learner committed:
Fix a rare flakyness because of faker.commerce.department list being just 20 items

so there is 5% percent chance of same thing getting returned twice and can cause already present error

yash-learner committed:
Merge branch 'issues/14296/add-tests-for-service-request' of https://github.com/ohcnetwork/care_fe into issues/14296/add-tests-for-service-request

yash-learner committed:
Merge branch 'develop' into issues/14296/add-tests-for-service-request
